src/mds/StrayManager.cc

```diff
@@ -0,0 +1,863 @@
+   * expiry code from seeing a dirty parent and trying to write a backtrace */
+  if (!trunc) {
+    if (in->is_dirty_parent()) {
+      in->clear_dirty_parent();
+    }
+  }
+
+
+  // Try to purge immediately if possible, else enqueue
+  const uint32_t ops_required = _calculate_ops_required(in, trunc);
+
+  bool consumed = _consume(dn, trunc, ops_required);
+  if (consumed) {
+    dout(10) << __func__ << ": purging this dentry immediately: "
+      << *dn << dendl;
+  } else {
+    dout(10) << __func__ << ": enqueuing this dentry for later purge: "
+      << *dn << dendl;
+    ready_for_purge.push_back(QueuedStray(dn, trunc, ops_required));
+  }
+}
+
+void StrayManager::_advance()
+{
+  std::list<QueuedStray>::iterator i;
+  for (i = ready_for_purge.begin();
+       i != ready_for_purge.end(); ++i) {
+    const QueuedStray &qs = *i;
+    const bool consumed = _consume(qs.dn, qs.trunc, qs.ops_required);
+    if (!consumed) {
+      break;
+    }
+  }
+
+  // Erase all the ones that returned true from _consume
+  ready_for_purge.erase(ready_for_purge.begin(), i);
+}
+
+/*
+ * Note that there are compromises to how throttling
+ * is implemented here, in the interests of simplicity:
+ *  * If insufficient ops are available to execute
+ *    the next item on the queue, we block even if
+ *    there are items further down the queue requiring
+ *    fewer ops which might be executable
+ *  * The ops considered "in use" by a purge will be
+ *    an overestimate over the period of execution, as
+ *    we count filer_max_purge_ops and ops for old backtraces
+ *    as in use throughout, even though towards the end
+ *    of the purge the actual ops in flight will be
+ *    lower.
+ *  * The ops limit may be exceeded if the number of ops
+ *    required by a single inode is greater than the
+ *    limit, for example directories with very many
+ *    fragments.
+ */
+bool StrayManager::_consume(CDentry *dn, bool trunc, uint32_t ops_required)
+{
+  const int files_avail = g_conf->mds_max_purge_files - files_purging;
+
+  if (files_avail <= 0) {
+    dout(20) << __func__ << ": throttling on max files" << dendl;
+    return false;
+  } else {
+    dout(20) << __func__ << ": purging dn: " << *dn << dendl;
+  }
+
+  // Calculate how much of the ops allowance is available, allowing
+  // for the case where the limit is currently being exceeded.
+  uint32_t ops_avail;
+  if (ops_in_flight <= max_purge_ops) {
+    ops_avail = max_purge_ops - ops_in_flight;
+  } else {
+    ops_avail = 0;
+  }
+
+  /* The ops_in_flight > 0 condition here handles the case where the
+   * ops required by this inode would never fit in the limit: we wait
+   * instead until nothing else is running */
+  if (ops_in_flight > 0 && ops_avail < ops_required) {
+    dout(20) << __func__ << ": throttling on max ops (require "
+             << ops_required << ", " << ops_in_flight << " in flight" << dendl;
+    return false;
+  }
```

Comment:
In this location, we wave through oversized jobs (which would otherwise block because ops_avail < ops_required) if nothing else is in flight.  Once we've started running that oversized job, its ops_required will be added to ops_in_flight in the usual way, so the next time through _consume we'll find ops_avail is 0 and ops_in_flight is > 0, so the next job will not be consumed.